### PR TITLE
pulsar updates...

### PR DESCRIPTION
**Pull request by daveads**
Refactor PulsarBackend for multi-channel support and improved error handling 
- Implement separate producers and consumers for each channel
- Add channel-specific subscribe and unsubscribe functionality
- Improve logging with a dedicated logger
- Enhance error handling and timeout management in next_published method
- Optimize connection and disconnection processes

### CODE REVIEW

File: broadcaster/_backends/pulsar.py [hunk 1]
```diff
@@ -14,64 +15,80 @@ def __init__(self, url: str):
...
-            logging.error(e)
+            logger.error(f"Error connecting to Pulsar: {e}")
             raise e
 
     async def disconnect(self) -> None:
-        if self._producer:
-            await anyio.to_thread.run_sync(self._producer.close)
-        if self._consumer:
-            await anyio.to_thread.run_sync(self._consumer.close)
+        for producer in self._producers.values():
```
Comment (danyi1212) [inline broadcaster/_backends/pulsar.py:34]: Can it be done concurrently using `await asyncio.gather()`?

File: broadcaster/_backends/pulsar.py [hunk 2]
```diff
@@ -14,64 +15,80 @@ def __init__(self, url: str):
         self._port = parsed_url.port or 6650
         self._service_url = f"pulsar://{self._host}:{self._port}"
         self._client = None
-        self._producer = None
-        self._consumer = None
+        self._producers = {}
+        self._consumers = {}
+        self._subscribed_channels = set()
```
Comment (danyi1212) [inline broadcaster/_backends/pulsar.py:20]: Isn't it exactly the same as `self._consumers.keys()`? Do we really need to store it twice?

File: broadcaster/_backends/pulsar.py [hunk 3]
```diff
@@ -14,64 +15,80 @@ def __init__(self, url: str):
...
+            self._consumers[channel] = consumer
+            logger.info(f"Subscribed to channel: {channel}")
 
     async def unsubscribe(self, channel: str) -> None:
-        # Similarly, we don't need to do anything here
-        pass
+        if channel in self._subscribed_channels:
+            self._subscribed_channels.remove(channel)
+            consumer = self._consumers.pop(channel, None)
+            if consumer:
```
Comment (danyi1212) [inline broadcaster/_backends/pulsar.py:58]: Better to first close the consumer then remove it from the dict. If an error is raised attempting to close, we will not have access to the channel anymore

File: broadcaster/_backends/pulsar.py [hunk 4]
```diff
@@ -14,64 +15,80 @@ def __init__(self, url: str):
...
-                msg = await anyio.to_thread.run_sync(self._consumer.receive)
-                channel, content = msg.data().decode("utf-8").split(":", 1)
-                await anyio.to_thread.run_sync(lambda: self._consumer.acknowledge(msg))
-                return Event(channel=channel, message=content)
-
-            except anyio.get_cancelled_exc_class():
-                # cancellation
-                logging.info("next_published task is being cancelled")
-                raise
+            if not self._consumers:
```
Comment (danyi1212) [inline broadcaster/_backends/pulsar.py:73]: Better to create a background task for each consumer that receives events from a specific channel, then make them all push into a shared queue and make the `next_published` get from that shared queue.
This way, receivers are tied to consumers and the next published get simplified.

Make sure to limit the queue size (to make the receivers wait on pushing) and terminate receivers on disconnect / unsubscribe.

**Review by danyi1212 — CHANGES_REQUESTED** [review]

File: broadcaster/_backends/pulsar.py (hunk 1 above)
Comment (daveads) [reply broadcaster/_backends/pulsar.py:34]: python 3.8 doesn't support asyncio.gather()...

**Comment by gemanor** [comment]
@daveads please address @danyi1212's comments

**Comment by daveads** [comment]
> @daveads please address @danyi1212's comments

Sure

**Comment by gemanor** [comment]
Hey @daveads any news on this?

**Comment by daveads** [comment]
> Hey @daveads any news on this?

@gemanor, I’ve been busy with OpenFGA integration and other tasks. I’ll send a commit tomorrow with the required changes.

**Comment by daveads** [comment]
@gemanor  @danyi1212  Done

**Comment by gemanor** [comment]
@danyi1212 it's LGTM, please run a last sanity review and merge.

File: broadcaster/_backends/pulsar.py [hunk 5]
```diff
@@ -1,77 +1,119 @@
...
-        pass
+        if channel in self._consumers:
+            consumer = self._consumers.pop(channel)
+            try:
+                await anyio.to_thread.run_sync(consumer.close)
+            except ValueError:
+                logger.warning(f"Consumer for channel {channel} was not in the client's list")
+            except Exception as e:
+                logger.error(f"Error closing consumer for channel {channel}: {e}")
+            logger.info(f"Unsubscribed from channel: {channel}")
```
Comment (danyi1212) [inline broadcaster/_backends/pulsar.py:90]: This should be in the `else` of that `try..except..else` clause

File: broadcaster/_backends/pulsar.py [hunk 6]
```diff
@@ -1,77 +1,119 @@
...
+            for producer in self._producers.values()
+        ] + [
+            anyio.to_thread.run_sync(consumer.close)
+            for consumer in self._consumers.values()
+        ]
+        
+        # Add client close coroutine if client exists
         if self._client:
-            await anyio.to_thread.run_sync(self._client.close)
+            close_coros.append(anyio.to_thread.run_sync(self._client.close))
```
Comment (danyi1212) [inline broadcaster/_backends/pulsar.py:54]: The order probably do matter here. The `_client.close()` should be after the producers and consumers are closed

File: broadcaster/_backends/pulsar.py [hunk 7]
```diff
@@ -1,77 +1,119 @@
...
+                    content = msg.data().decode("utf-8")
+                    await anyio.to_thread.run_sync(consumer.acknowledge, msg)
+                    await self._shared_queue.put(Event(channel=channel, message=content))
+                    logger.info(f"Received message from channel {channel}: {content}")
+                except Exception as e:
+                    logger.error(f"Error receiving message from channel {channel}: {e}")
+        except asyncio.CancelledError:
+            logger.info(f"Receiver for channel {channel} was cancelled")
+        finally:
+            await anyio.to_thread.run_sync(consumer.close)
```
Comment (danyi1212) [inline broadcaster/_backends/pulsar.py:119]: Closing the consumer is done also in the disconnect. That will cause consumers to be closed twice, is it fine from the Pulsar's consumer interface?

File: tests/test_broadcast.py [hunk 8]
```diff
@@ -22,16 +23,18 @@ async def test_redis():
             assert event.channel == "chatroom"
             assert event.message == "hello"
 
-## pulsar test 
+
 @pytest.mark.asyncio
 async def test_pulsar():
     async with Broadcast("pulsar://localhost:6650") as broadcast:
         async with broadcast.subscribe("chatroom") as subscriber:
+            await asyncio.sleep(2)  # Increase this value
```
Comment (danyi1212) [inline tests/test_broadcast.py:31]: Why is this sleep needed?

**Review by danyi1212 — CHANGES_REQUESTED** [review]

File: broadcaster/_backends/pulsar.py [hunk 9]
```diff
@@ -1,77 +1,119 @@
...
+        if channel not in self._consumers:
+            consumer = await anyio.to_thread.run_sync(
+                lambda: self._client.subscribe(
+                    channel,
+                    subscription_name=f"broadcast_subscription_{channel}",
+                    consumer_type=pulsar.ConsumerType.Shared,
+                )
+            )
+            self._consumers[channel] = consumer
+            self._receiver_tasks[channel] = asyncio.create_task(self._receiver(channel, consumer))
```
Comment (danyi1212) [inline broadcaster/_backends/pulsar.py:77]: There is probably a need to do a quick `await asyncio.sleep(0)` here to allow the event loop to briefly begin the new coroutine.

**Comment by daveads** [comment]
will work on the review this week...

File: broadcaster/_backends/pulsar.py (hunk 7 above)
Comment (daveads) [reply broadcaster/_backends/pulsar.py:119]: it's actually fine for consumers to be closed both in the receiver task's cleanup and in the disconnect() method.

File: tests/test_broadcast.py (hunk 8 above)
Comment (daveads) [reply tests/test_broadcast.py:31]: can't fully remember... buh i knew pulsar take time to boot up...

**Comment by daveads** [comment]
we were suppose to drop test for 3.8 and add 3.12.. 

will get back to this... need to attend to something via opal...

**Comment by daveads** [comment]
Replaced anyio.to_thread.run_sync() with asyncio.to_thread()

**Review by danyi1212 — CHANGES_REQUESTED** [review]
Looks very good, left few small comments

**Comment by danyi1212** [comment]
Notice the difference between 
```python
await asyncio.gather(*coros, return_exceptions=True)  
```
Compared to
```python
try:
    await asyncio.gather(*coros)
except Exception as e:
    logger.exception("Failed ...")
```

In the first option it will supress any exception raised inside the coros and include them in the list returned by the gather, retaining the order of the received coros.

In the second example, when a single coro inside the gather fails, it will immediately cancel all running coros, and drop the returned values of completed coros.

If you want all coros to complete regardless of when some of them fail, you should use return_exceptions=True. Through you should remember to actually use the returned values, and if exception where raised, handle them (in our case, at least log them)

**Review by danyi1212 — APPROVED** [review]

**Pull request merged**